### PR TITLE
Eip 4399

### PR DESCRIPTION
**Pull request by jflo**
## PR description

Implements EIP-4399 by repurposing the DIFFICULTY opcode to provide entropy provided by beacon chain post merge.

This is intended to change the behavior during the merge, and leaving the semantic changes to clarify this behavior till after the merge, when the opcode can be documented as renamed to RANDOM, thought its instruction byte remains the same.

## Changelog

- [X] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

**Review by garyschulte — COMMENTED** [review]
I was expecting a new RandomOperation with the same opcode and/or changes to DifficultyOperation to[rename it](https://github.com/ethereum/EIPs/blob/979cf123558bab61e2278f3c962c9a68d1adc2f4/EIPS/eip-4399.md#renaming)

I don't have a great suggestion on how to do this renaming on the basis of a non-protocolSpec change.  

I don't think there is a functional problem with not renaming or aliasing the operation, other than that traces might not show the 'correct' operation name.  

If indeed there is no functional difference between the op name DIFFICULTY and RANDOM, we could make the name change part of the EIP-4488 hardfork 🤔

File: evm/src/main/java/org/hyperledger/besu/evm/operation/DifficultyOperation.java
Comment (shemnon) [inline evm/src/main/java/org/hyperledger/besu/evm/operation/DifficultyOperation.java:32]: Don't make this operation do two different things.  Make a separate `RandomOperation` and have the hard fork spec drop it in.

**Review by shemnon — CHANGES_REQUESTED** [review]

File: evm/src/main/java/org/hyperledger/besu/evm/operation/DifficultyOperation.java
Comment (jflo) [reply evm/src/main/java/org/hyperledger/besu/evm/operation/DifficultyOperation.java:32]: I wanted to do that, but since they are the same byte, didn't think it was possible. If you're saying that dropping it in after Difficulty will override it, then that is much cleaner. Will do.

Comment (jflo) [reply evm/src/main/java/org/hyperledger/besu/evm/operation/DifficultyOperation.java:32]: Ah now I remember. I didn't do that either because this isn't driven by a protocol schedule. This happens when difficulty becomes zero on the merge. See the definition of `TERMINAL_BLOCK` here https://eips.ethereum.org/EIPS/eip-3675#definitions

Comment (jflo) [reply evm/src/main/java/org/hyperledger/besu/evm/operation/DifficultyOperation.java:32]: Ok, so what we could do, is drop in RANDOM to overwrite DIFFICULTY for 0x44, at some hardfork block before the merge. The conditional has to stay in RANDOM though, since it shouldn't change behavior till difficulty value disappears from block. Then we clean it up in the first release after the merge.

Comment (garyschulte) [reply evm/src/main/java/org/hyperledger/besu/evm/operation/DifficultyOperation.java:32]: I think this is the most viable option, given the merge transition mechanism.  If an EIP-4488 hardfork does not happen pre-merge, we could add it to the soft fork that is (was?) slated to happen a couple weeks prior to the merge which adds the target terminal total difficulty.

File: evm/src/main/java/org/hyperledger/besu/evm/operation/DifficultyOperation.java
Comment (garyschulte) [inline evm/src/main/java/org/hyperledger/besu/evm/operation/DifficultyOperation.java:21]: unused import?

File: evm/src/main/java/org/hyperledger/besu/evm/operation/RandomOperation.java
Comment (garyschulte) [inline evm/src/main/java/org/hyperledger/besu/evm/operation/RandomOperation.java:22]: 👍

File: evm/src/main/java/org/hyperledger/besu/evm/MainnetEVMs.java
Comment (garyschulte) [inline evm/src/main/java/org/hyperledger/besu/evm/MainnetEVMs.java:372]: LGTM, perhaps once the EIP-4488 hardfork has a name, we can rename this EVM

**Review by garyschulte — APPROVED** [review]
LGTM, as long as spotless doesn't complain about the unused import

File: besu/src/main/java/org/hyperledger/besu/cli/config/NetworkName.java
Comment (shemnon) [inline besu/src/main/java/org/hyperledger/besu/cli/config/NetworkName.java:24]: I don't think this is the type of network we should be putting in our named networks.

**Review by shemnon — CHANGES_REQUESTED** [review]
I'm not liking putting a test network in the list of named networks.  Testing would be better served by using the `--genesis-file=preMerge.json` CLI flag.

File: config/src/main/resources/preMerge.json
Comment (jflo) [reply config/src/main/resources/preMerge.json]: moved this to test

**Review by shemnon — APPROVED** [review]
Is philly an official name or a filly-in name? ;)

**Pull request merged**
